### PR TITLE
add training optimizations

### PR DESCRIPTION
**Pull request by synchtp**
- Enable TF32.
- Use flash attention.
- Use Bf16, not FP16 -> more stable training, no need for gradient scaler.

### CODE REVIEW

File: miner.py [hunk 1]
```diff
@@ -393,11 +406,12 @@ def download_file( bucket, filename ):
     parser.add_argument('--netuid', type=int, default=212, help='Bittensor network UID.')
     parser.add_argument('--bucket', type=str, default='decis', help='S3 bucket name')
     parser.add_argument('--desired_batch_size', type=int, default=512, help='Training batch size per step')
-    parser.add_argument('--actual_batch_size', type=int, default=9, help='Training batch size per accumulation.')
+    parser.add_argument('--actual_batch_size', type=int, default=8, help='Training batch size per accumulation.')
```
Comment (synchtp) [inline miner.py:409]: @unconst was this a typo or intended to b 9?

File: miner.py [hunk 2]
```diff
@@ -272,33 +281,37 @@ def download_file( bucket, filename ):
...
-            optimizer.zero_grad()
             total_loss = 0.0
             total_steps = config.desired_batch_size // config.actual_batch_size
             progress_bar = tqdm(total=total_steps, desc="Training:")
             for idx, batch in enumerate(dataset):
                 input_ids = torch.tensor(batch, dtype=torch.long).to(model.device)
                 labels = input_ids.clone()
                 labels = torch.where(labels == hparams.tokenizer.pad_token_id, -100, labels)
-                with torch.amp.autocast( device_type = model.device.type, dtype = torch.float16 ):  # Enable autocasting for mixed precision
+                with torch.amp.autocast( device_type = model.device.type, dtype = torch.bfloat16 ):  # Enable autocasting for mixed precision
```
Comment (synchtp) [inline miner.py:298]: In LLM training, almost every shifted from FP16 -> BF16. It has the same range as FP32, so better numerical stability.

**Review by unconst — APPROVED** [review]
